crates/services/sync/src/import/cache.rs

```diff
@@ -0,0 +1,533 @@
+                    None,
+                    height..height.saturating_add(1),
+                    vec![data],
+                ))
+            }
+            (CachedDataBatch::None(_), CachedData::Block(data)) => {
+                CachedDataBatch::Blocks(Batch::new(
+                    None,
+                    height..height.saturating_add(1),
+                    vec![data],
+                ))
+            }
+            (CachedDataBatch::Headers(mut batch), CachedData::Header(data)) => {
+                batch.range = batch.range.start..height.saturating_add(1);
+                batch.results.push(data);
+                CachedDataBatch::Headers(batch)
+            }
+            (CachedDataBatch::Blocks(mut batch), CachedData::Block(data)) => {
+                batch.range = batch.range.start..height.saturating_add(1);
+                batch.results.push(data);
+                CachedDataBatch::Blocks(batch)
+            }
+            (CachedDataBatch::Headers(headers_batch), CachedData::Block(block)) => {
+                chunks.push(CachedDataBatch::Headers(headers_batch));
+                CachedDataBatch::Blocks(Batch::new(
+                    None,
+                    height..height.saturating_add(1),
+                    vec![block],
+                ))
+            }
+            (CachedDataBatch::Blocks(blocks_batch), CachedData::Header(header)) => {
+                chunks.push(CachedDataBatch::Blocks(blocks_batch));
+                CachedDataBatch::Headers(Batch::new(
+                    None,
+                    height..height.saturating_add(1),
+                    vec![header],
+                ))
+            }
+        }
+    }
+
+    fn check_chunk_limit(
+        &self,
+        current_chunk: CachedDataBatch,
+        max_chunk_size: usize,
+        chunks: &mut Vec<CachedDataBatch>,
+        current_height: u32,
+    ) -> CachedDataBatch {
+        match current_chunk {
+            CachedDataBatch::Headers(batch) => {
+                if batch.results.len() >= max_chunk_size {
+                    chunks.push(CachedDataBatch::Headers(batch));
+                    CachedDataBatch::None(current_height..current_height)
+                } else {
+                    CachedDataBatch::Headers(batch)
+                }
+            }
+            CachedDataBatch::Blocks(batch) => {
+                if batch.results.len() >= max_chunk_size {
+                    chunks.push(CachedDataBatch::Blocks(batch));
+                    CachedDataBatch::None(current_height..current_height)
+                } else {
+                    CachedDataBatch::Blocks(batch)
+                }
+            }
+            CachedDataBatch::None(range) => CachedDataBatch::None(range),
+        }
+    }
+
+    fn push_current_chunk(
+        &self,
+        chunks: &mut Vec<CachedDataBatch>,
+        current_chunk: CachedDataBatch,
+    ) {
+        if !current_chunk.is_range_empty() {
+            chunks.push(current_chunk);
+        }
```

Comment:
I think moving it to a separate function is overkill, it would be easier to read it in the loop and after the loop, because it is logical check and pretty simple=)